app/controllers/application_controller.rb

```diff
@@ -929,6 +930,29 @@ def should_enforce_2fa?
       !current_user.has_any_second_factor_methods_enabled?
   end
 
+  def redirect_to_profile_if_required
+    return if request.format.json? && is_api?
```

Comment:
I realised that we actually need to just `return if request.format.json?` here, if you take another look at `should_enforce_2fa?` we actually do not do the redirect if the request is JSON which makes sense:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/controllers/application_controller.rb#L922-L923

It's confusing because it's mixed up with the "redirect to login" logic:

https://github.com/discourse/discourse/blob/a1e330e6504988ed24d4b792215b717f4484c34a/app/controllers/application_controller.rb#L878-L879

This is why it will be excellent when you refactor this so the whole "redirect user if we want to force them to do something" logic will be in the same place :)

For now we should just fix this case and add a request spec to catch it. The steps I followed to find the issue:

1. Make a required for all field
2. As an admin, visit /admin route manually in my browser address bar
3. Click the "Back to forum" button in the sidebar
4. Get errors in the topic list, because it cannot handle a 302 response from the server

At the point of 4, I think we actually need a top-level `willTransition` check in our application.js route:

```
@action
willTransition(transition) {
  if (this.currentUser.needs_required_fields_check && transition.to?.name != "preferences.profile") {
      transition.abort();
      return this.router.transitionTo("preferences.profile");
  }

  return true;
},
```

This will be helpful for the 2FA route, I am guessing a similar loophole exists there now.

Reply:
> At the point of 4, I think we actually need a top-level willTransition check in our application.js route

Yes. I have a separate issue opened for this relating to 2FA, where you can go to the `/faq` page and then use client-side routing from there. I have a fix coming up for that. It should cover this use case as well. 🙏